examples/pytorch/graphsage/multi_gpu_node_classification.py

```diff
@@ -1,182 +1,164 @@
-        t0 = time.time()
+        total_loss = 0
         for it, (input_nodes, output_nodes, blocks) in enumerate(train_dataloader):
             x = blocks[0].srcdata['feat']
-            y = blocks[-1].dstdata['label'][:, 0]
+            y = blocks[-1].dstdata['label']
             y_hat = model(blocks, x)
             loss = F.cross_entropy(y_hat, y)
             opt.zero_grad()
             loss.backward()
             opt.step()
-            if it % 20 == 0 and rank == 0:
-                acc = MF.accuracy(y_hat, y)
-                mem = torch.cuda.max_memory_allocated() / 1000000
-                print('Loss', loss.item(), 'Acc', acc.item(), 'GPU Mem', mem, 'MB')
-        tt = time.time()
-
-        if rank == 0:
-            print(tt - t0)
-        durations.append(tt - t0)
-
-        model.eval()
-        ys = []
-        y_hats = []
-        for it, (input_nodes, output_nodes, blocks) in enumerate(valid_dataloader):
-            with torch.no_grad():
-                x = blocks[0].srcdata['feat']
-                ys.append(blocks[-1].dstdata['label'])
-                y_hats.append(model.module(blocks, x))
-        acc = MF.accuracy(torch.cat(y_hats), torch.cat(ys)) / world_size
+            total_loss += loss
+        acc = evaluate(model, g, val_dataloader).to(device) / nprocs
         dist.reduce(acc, 0)
-        if rank == 0:
-            print('Validation acc:', acc.item())
-        dist.barrier()
-
-    if rank == 0:
-        print(np.mean(durations[4:]), np.std(durations[4:]))
-    model.eval()
-    with torch.no_grad():
-        # since we do 1-layer at a time, use a very large batch size
-        pred = model.module.inference(graph, device='cuda', batch_size=2**16)
-        if rank == 0:
-            acc = MF.accuracy(pred[test_idx], graph.ndata['label'][test_idx])
-            print('Test acc:', acc.item())
+        if (proc_id == 0):
+            print("Epoch {:05d} | Loss {:.4f} | Accuracy {:.4f} "
+                  .format(epoch, total_loss / (it+1), acc.item()))
+
+def run(proc_id, nprocs, devices, g, dist_data, mode):
+    # find corresponding device for my rank
+    device = devices[proc_id]
+    torch.cuda.set_device(device)
+    # initialize process group and unpack data for sub-processes
+    dist.init_process_group(backend="nccl", init_method='tcp://127.0.0.1:12345', world_size=nprocs, rank=proc_id)
+    out_size, train_idx, val_idx, test_idx = dist_data
+    train_idx = train_idx.to(device)
+    val_idx = val_idx.to(device)
+    g = g.to(device if mode == 'puregpu' else 'cpu')
+    # create RGCN model (distributed)
+    in_size = g.ndata['feat'].shape[1]
+    model = SAGE(in_size, 256, out_size).to(device)
+    model = DistributedDataParallel(model, device_ids=[device], output_device=device)
+    # training + testing
+    use_uva = (mode == 'mixed')
+    train(proc_id, nprocs, device, g, train_idx, val_idx, model, use_uva)
+    layerwise_infer(proc_id, device, g, test_idx, model, use_uva, batch_size=2**16)    
```

Comment:
I'll prefer directly setting the default value of `batch_size` in defining `layerwise_infer`.

Reply:
Updated accordingly